helix-core/src/main/java/org/apache/helix/controller/rebalancer/strategy/StickyRebalanceStrategy.java

```diff
@@ -0,0 +1,206 @@
+      final List<String> liveNodes, final Map<String, Map<String, String>> currentMapping,
+      ResourceControllerDataProvider clusterData) {
+    ZNRecord znRecord = new ZNRecord(_resourceName);
+    if (liveNodes.isEmpty()) {
+      return znRecord;
+    }
+
+    if (clusterData.getSimpleCapacitySet() == null) {
+      logger.warn("No capacity set for resource: {}", _resourceName);
+      return znRecord;
+    }
+
+    // Sort the assignable nodes by id
+    List<CapacityNode> assignableNodes = new ArrayList<>(clusterData.getSimpleCapacitySet());
+    assignableNodes.sort(Comparator.comparing(CapacityNode::getId));
+
+    // Filter out the nodes if not in the liveNodes parameter
+    // Note the liveNodes parameter here might be processed within the rebalancer, e.g. filter based on tags
+    Set<String> liveNodesSet = new HashSet<>(liveNodes);
+    assignableNodes.removeIf(n -> !liveNodesSet.contains(n.getId()));
+
+    //  Populate valid state map given current mapping
+    Map<String, Map<String, String>> stateMap =
+        populateValidStateMapFromCurrentMapping(currentMapping, assignableNodes);
+
+    if (logger.isDebugEnabled()) {
+      logger.debug("currentMapping: {}", currentMapping);
+      logger.debug("stateMap: {}", stateMap);
+    }
+
+    // Assign partitions to node by order.
+    for (int i = 0, index = 0; i < _partitions.size(); i++) {
+      int startIndex = index;
+      for (Map.Entry<String, Integer> entry : _states.entrySet()) {
+        String state = entry.getKey();
+        int stateReplicaNumber = entry.getValue();
+        // For this partition, compute existing number replicas
+        long existsReplicas =
+            stateMap.computeIfAbsent(_partitions.get(i), m -> new HashMap<>()).values().stream()
+                .filter(s -> s.equals(state)).count();
+        for (int j = 0; j < stateReplicaNumber - existsReplicas; j++) {
+          while (index - startIndex < assignableNodes.size()) {
+            CapacityNode node = assignableNodes.get(index++ % assignableNodes.size());
+            if (node.canAdd(_resourceName, _partitions.get(i))) {
+              stateMap.get(_partitions.get(i)).put(node.getId(), state);
+              break;
+            }
+          }
+
+          if (index - startIndex >= assignableNodes.size()) {
+            // If the all nodes have been tried out, then no node can be assigned.
+            logger.warn("No enough assignable nodes for resource: {}", _resourceName);
+          }
+        }
+      }
+    }
+    for (Map.Entry<String, Map<String, String>> entry : stateMap.entrySet()) {
+      znRecord.setListField(entry.getKey(), new ArrayList<>(entry.getValue().keySet()));
+    }
+    if (logger.isDebugEnabled()) {
+      logger.debug("znRecord: {}", znRecord);
+    }
+
+    return znRecord;
+  }
+
+  private int countStateReplicas() {
+    int total = 0;
+    for (Integer count : _states.values()) {
+      total += count;
+    }
+    return total;
+  }
```

Comment:
So far there is no _states update. Let's just have a variable to memorize the state count instead of keep recomputing it.